payment/queue.py

```diff
@@ -1,234 +1,199 @@
+            app_id=value.app_id,
+            objekt='payment',
+            state='success',
+            action='send' if wallet_address == value.sender_address else 'receive',
+            value=value.to_primitive())
+
+    def enqueue_payment_failed_callback(self, request: PaymentRequest, reason: str):
+        self.__enqueue_callback(
+            callback=request.callback,
+            app_id=request.app_id,
+            objekt='payment',
+            state='fail',
+            action='send',
+            value={'id': request.id, 'reason': reason})
+
+    def enqueue_wallet_failed_callback(self, request: WalletRequest, reason: str):
+        self.__enqueue_callback(
+            callback=request.callback,
+            app_id=request.app_id,
+            objekt='wallet',
+            state='fail',
+            action='create',
+            value={'id': request.id, 'reason': reason})
+
+    async def call_callback(self, callback: str, app_id: str, objekt: str, state: str, action: str, value: dict):
+
+        async def retry_callback(callback: str, payload: dict):
+            for i in range(5):
+                try:
+                    # The kin client sets x-www-form-urlencoded as the content type, so override it just for this request
+                    async with self.session.post(callback, json=payload, raise_for_status=True,
+                                                 headers={'Content-Type': 'application/json'}) as resp:
+                        return await resp.json()
+                except:
+                    if i == 4:
+                        raise
+
+                await asyncio.sleep(0.2)
+
+        payload = {
+            'object': objekt,
+            'state': state,
+            'action': action,
+            'value': value,
+        }
+
+        tags = ['app_id:%s' % app_id,
+                'object:%s' % objekt,
+                'state:%s' % state,
+                'action:%s' % action]
+        try:
+            response = await retry_callback(callback, payload)
+            log.info('callback response', response=response, payload=payload)
+            statsd.increment('callback.success', tags=tags)
+        except Exception as e:
+            log.error('callback failed', error=e, payload=payload)
+            statsd.increment('callback.failed', tags=tags)
 
-    else:
-        statsd.increment('wallet.created', tags=['app_id:%s' % wallet_request.app_id])
-        wallet = get_wallet(wallet_request.wallet_address)
-        wallet.id = wallet_request.id  # XXX id is required in webhook
-        enqueue_wallet_callback(wallet_request, wallet)
+    async def pay_and_callback(self, payment_request: PaymentRequest):
+        """lock, try to pay and callback."""
+        log.info('pay_and_callback received', payment_request=payment_request)
+        try:
+            async with RedisLock(self.redis_conn, '_lock_payment:{}'.format(payment_request.id),
+                                 timeout=60, wait_timeout=0):
```

Comment:
when setting `wait_timeout`, I think you need to test that you actually received the lock: `lock.is_owner`

Reply:
wait_timeout 0 causes it to raise an exception if it can't get the lock immediately, it won't just go in the with block without the lock